utils/payment.js

```
@@ -0,0 +1,188 @@
+const FAILURE = 'failure';
+const SUCCESS = 'success';
+const METHOD_APPLE_PAY = 'apple';
+const SELECTOR_APPLE_PAY = '.apple-pay-button';
+const CLASS_HIDDEN = 'o-normalise-visually-hidden';
+
+/**
+ * Validate ourselves with Apple through Membership and get a payment session token
+ * @param {string} url Merchant validation URL
+ * @return {promise<string>} session token
+ */
+async function fetchPaymentSession (url) {
+	// Call Membership with the URL so they can validate session
+	const sessionToken = await fetch('membership', url);
+	return sessionToken;
+}
+
+/**
+ * Process the response from Apple Pay with Membership
+ * @param {object} response Apple Pay response
+ * @return {promise<string>} `success` or `failure`
+ */
+async function processResponse (response) {
+	// Call Membership with the response so they can return a status
+	try {
+		const status = await fetch('membership', response);
+		return status ? SUCCESS : FAILURE;
+	} catch (e) {
+		return FAILURE;
+	}
+}
+
+/**
+ * Payment wrapper for Payment Request API interaction
+ * @example
+ * // Create a payment instance
+ * const payment = new Payment(window, Payment.METHOD_APPLE_PAY, {value: 1.00, currency: 'GBP', label: 'Standard FT'});
+ * const shown = await payment.showPaymentButton();
+ * if (!shown) {
+ *     // Browser doesn't support the payment method
+ * }
+ */
+class Payment {
+	/**
+	 * Check the Payment Request API is available and the methods button is available
+	 * @param {Window} window
+	 * @param {string} method
+	 * @param {number} value Decimal price
+	 * @param {string} currency ISO 3 digit currency code
+	 * @param {string} label Name for what's being sold
+	 * @throws If the payment button is not found
+	 */
+	constructor (window, method, {value, currency, label}) {
+		if (!window.PaymentRequest) {
+			// For method='apple' fallback to using Apple Pay JS instead?
+			return;
```

I reckon we should throw an Error if PaymentRequest isn't present